src/bm_sim/lpm_trie.cpp

```diff
@@ -0,0 +1,294 @@
+}
+
+
+LPMTrie::LPMTrie(std::size_t key_width_bytes) :
+    key_width_bytes(key_width_bytes) {
+    assert(key_width_bytes <= 64);
+    root = std::make_unique<Node>();
+  }
+
+LPMTrie::LPMTrie(LPMTrie &&other) noexcept :
+    key_width_bytes(other.key_width_bytes) {
+  root.reset(other.root.release());
+}
+
+LPMTrie::~LPMTrie() {}
+
+LPMTrie &LPMTrie::operator=(LPMTrie &&other) noexcept {
+  key_width_bytes = other.key_width_bytes;
+  root.reset(other.root.release());
+  return *this;
+}
+
+void LPMTrie::insert(const std::string &prefix, int prefix_length,
+                     value_t value) {
+  Node *current_node = root.get();
+  byte_t byte;
+  for (int i = 0; prefix_length >= 8; ++i) {
+    byte = static_cast<byte_t>(prefix[i]);
+    Node *node = current_node->get_next_node(byte);
+
+    if (!node) {
+      auto new_node = std::make_unique<Node>(current_node, byte);
+      node = new_node.get();
+      current_node->set_next_node(byte, std::move(new_node));
+    }
+
+    prefix_length -= 8;
+    current_node = node;
+  }
+
+  byte_t key = static_cast<byte_t>(prefix.back()) >> (8 - prefix_length);
+  current_node->insert_prefix(prefix_length, key, value);
+}
+
+bool LPMTrie::retrieve_value(const std::string &prefix, int prefix_length,
+                             value_t *value) const {
+  Node *current_node = root.get();
+  byte_t byte;
+
+  for (int i = 0; prefix_length >= 8; ++i) {
+    byte = static_cast<byte_t>(prefix[i]);
+    current_node = current_node->get_next_node(byte);
+    if (!current_node)
+      return false;
+    prefix_length -= 8;
+  }
+
+  byte_t key = static_cast<byte_t>(prefix.back()) >> (8 - prefix_length);
+  return current_node->get_prefix(prefix_length, key, value);
+}
+
+bool LPMTrie::has_prefix(const std::string &prefix, int prefix_length) const {
+  value_t value = 0;
+  return retrieve_value(prefix, prefix_length, &value);
+}
+
+bool LPMTrie::remove(const std::string &prefix, int prefix_length) {
+  Node *current_node = root.get();
+  byte_t byte;
+
+  for (int i = 0; prefix_length >= 8; ++i) {
+    byte = static_cast<byte_t>(prefix[i]);
+    current_node = current_node->get_next_node(byte);
+    if (!current_node)
+      return false;
+    prefix_length -= 8;
+  }
+
+  byte_t key = static_cast<byte_t>(prefix.back()) >> (8 - prefix_length);
+  if (!current_node->delete_prefix(prefix_length, key))
+    return false;
+
+  while (current_node->is_empty()) {
+    Node *tmp = current_node;
+    current_node = current_node->get_parent();
+    if (!current_node)
+      break;
+    current_node->delete_branch(tmp->get_child_id());
+  }
+
+  return true;
+}
+
+bool LPMTrie::lookup(const std::string &key, value_t *value) const {
+  Node *current_node = root.get();
+  byte_t byte;
+  size_t key_width = key_width_bytes;
+
+  int key_idx = 0;
+  while (current_node) {
+    if (key_width == 0) {
+      return current_node->get_empty_prefix(value);
+    }
+
+    for (auto &prefix : current_node->get_prefixes()) {
```

Comment:
I don't know if it's right to expose the internal `prefixes` using a `Node::get_prefixes` method. Do you think the functionality of this `for` loop should be moved to a dedicated `Node` method, given that this is the only usage?